### PR TITLE
add repetition penalty support

### PR DESCRIPTION
**Pull request by XiaobingSuper**
## Motivation

This PR is about adding repetition penalty support.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
@merrymercy, I re-added the repetition penalty that was removed by https://github.com/sgl-project/sglang/pull/3988. I found this issue when comparing the sglang output with HF output when the repetition penalty was applied. Please help review it. Thank you.

**Comment by merrymercy** [comment]
Why do you need this? OpenAI API does not provide this functionality. Why are the frequency and presence penalty not enough?

**Comment by XiaobingSuper** [comment]
@merrymercy , I do offline generation for one of my use cases, which wants to align with HF output, and another use case mainly uses Python Request. Do you mean if I use the frequency and presence penalty, I can get the same behavior with the repetition penalty method?

**Comment by merrymercy** [comment]
Can you disable repetition penalty for your HF use cases?

**Comment by XiaobingSuper** [comment]
> Can you disable repetition penalty for your HF use cases?

I can, but I think we shouldn't forbid this usage case for our customer users.

**Comment by merrymercy** [comment]
make sense. We can merge this. Can you add some test cases here? https://github.com/sgl-project/sglang/blob/094891c01a29ebef8bc1977f434b33283f9114c9/test/srt/test_penalty.py#L62-L70

**Comment by XiaobingSuper** [comment]
> make sense. We can merge this. Can you add some test cases here?
> 
> https://github.com/sgl-project/sglang/blob/094891c01a29ebef8bc1977f434b33283f9114c9/test/srt/test_penalty.py#L62-L70

Done.

**Comment by XiaobingSuper** [comment]
@merrymercy can we merge it?

**Comment by XiaobingSuper** [comment]
@merrymercy

**Comment by XiaobingSuper** [comment]
@merrymercy could you help review it? thanks!

**Comment by THU-LIJX** [comment]
@merrymercy We initially used vllm for model inference, and recently we plan to integrate sglang, allowing users to choose between vllm and sglang for inference. However, since we previously set the repetition_penalty parameter when using vllm for some tasks, and currently sglang does not support this parameter. It is difficult for us to align the results between vllm and sglang. We hope to have the repetition penalty feature reintegrated to sglang. Thanks!

**Comment by syskn** [comment]
Is this going to be merged? HF repetition penalty and OpenAI freq/presence penalties have highly different behavior (mainly because HF rep-pen takes the whole prefill context into account, while freq/pres. penalties are only taking currently generated tokens in account) and it is quite a pain having to manually merge this for every SGLang version for our use case.

**Comment by XiaobingSuper** [comment]
cc @merrymercy

**Comment by junliu-mde** [comment]
Support this PR. For small models or those without sufficient SFT, using "anyway" to prevent the model from repeating is still quite necessary.

**Comment by liguodongiot** [comment]
@XiaobingSuper Hi, Currently, repetition penalty only considers the generated text, not the original input text. Both HF Transformers and vLLM consider the original input text.

Modify as follows:
```
class BatchedRepetitionPenalizer(_BatchedPenalizer):

    def _prepare(self):
        batch_cumulated_repetition_penalties = []
        for req in self.orchestrator.reqs():
            cumulated_repetition_penalties_lst = [1] * self.orchestrator.vocab_size
            for idx in req.origin_input_ids:
                cumulated_repetition_penalties_lst[idx] = req.sampling_params.repetition_penalty
            batch_cumulated_repetition_penalties.append(cumulated_repetition_penalties_lst)

        self.cumulated_repetition_penalties = torch.tensor(
            data=batch_cumulated_repetition_penalties,
                dtype=torch.float32,
                device=self.orchestrator.device,
        )

        self.repetition_penalties = (
            torch.tensor(
                data=[
                    req.sampling_params.repetition_penalty
                    for req in self.orchestrator.reqs()
                ],
                dtype=torch.float32,
                device=self.orchestrator.device,
            )
        ).unsqueeze_(1)
```
